### PR TITLE
[hitrecord] Add new extractor (closes #10867)

### LINKED ISSUE #10867: Support hitrecord.org
## Please follow the guide below
- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your _issue_ (like that [x])
- Use _Preview_ tab to see how your issue will actually look like

---
### Make sure you are using the _latest_ version: run `youtube-dl --version` and ensure your version is _2016.10.02_. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [ ] I've **verified** and **I assure** that I'm running youtube-dl **2016.10.02**
### Before submitting an _issue_ make sure you have:
- [x ] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
### What is the purpose of your _issue_?
- [ ] Bug report (encountered problems with youtube-dl)
- [x] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---
### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your _issue_

---
### If the purpose of this _issue_ is a _site support request_ please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://hitrecord.org/records/2954362

---
### Description of your _issue_, suggested solution and other information

Looking to see how I can add or request support to add videos from hitrecord.org to be downloaded with `youtube-dl`! Please let me know if there's any further information required!

### PR DESCRIPTION
**Pull request by TRox1972**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

### CODE REVIEW

File: youtube_dl/extractor/hitrecord.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        video_id = self._match_id(url)
+        video_info = self._download_json(self._API_BASE + video_id, video_id)
+        user_info = video_info.get('user')
+
+        return {
+            'id': video_id,
+            'title': video_info['title'],
+            'url': video_info['source_url']['mp4_url'],
+            'description': clean_html(video_info.get('body')),
+            'uploader': user_info.get('username'),
```
Comment (dstftw) [inline youtube_dl/extractor/hitrecord.py]: `user_info` may be `None`.

File: youtube_dl/extractor/hitrecord.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    int_or_none,
+    unified_strdate,
+)
+from ..compat import compat_str
+
+
+class HitRecordIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?hitrecord\.org/records/(?P<id>\d+)'
+
+    _API_BASE = 'https://hitrecord.org/api/web/records/'
```
Comment (dstftw) [inline youtube_dl/extractor/hitrecord.py]: No point in separate variable.

File: youtube_dl/extractor/hitrecord.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        user_info = video_info.get('user')
+
+        return {
+            'id': video_id,
+            'title': video_info['title'],
+            'url': video_info['source_url']['mp4_url'],
+            'description': clean_html(video_info.get('body')),
+            'uploader': user_info.get('username'),
+            'uploader_id': compat_str(user_info.get('id')),
+            'release_date': unified_strdate(video_info.get('created_at')),
```
Comment (dstftw) [inline youtube_dl/extractor/hitrecord.py]: `created_at_i` as `timestamp`.

Comment (TRox1972) [inline youtube_dl/extractor/hitrecord.py]: when extracting timestamp, I get the error `﻿AssertionError: {'upload_date'} is not false : Missing keys in test definition: upload_date`.  Do you know how to fix this?

Comment (dstftw) [inline youtube_dl/extractor/hitrecord.py]: Add `upload_date` to test obviously.

Comment (TRox1972) [inline youtube_dl/extractor/hitrecord.py]: already tried that; got error `AssertionError: '20160818' != None : Invalid value for field uploade_date, expected '20160818', got None`

**Pull request closed without merging**